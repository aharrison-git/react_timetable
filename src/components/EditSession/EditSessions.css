.editform {
    text-align: left;
    display: grid;
    grid-template-columns: 120px 200px;
    grid-template-rows: repeat(4, 30px) 40px;
}

.cell {
    margin-left: 10px;
}

.row1 {
    grid-row-start: 1;
    grid-row-end: 1;
}

.row2 {
    grid-row-start: 2;
    grid-row-end: 2;
}

.row3 {
    grid-row-start: 3;
    grid-row-end: 3;
}

.row4 {
    grid-row-start: 4;
    grid-row-end: 4;
}

.row5 {
    grid-row-start: 5;
    grid-row-end: 5;
    width: 50%;
}

.column1 {
    grid-column-start: 1;
    grid-column-end: 1;
}

.column2 {
    grid-column-start: 2;
    grid-column-end: 2;
}

.success {
    color:  lightgreen;
    font-weight: normal;
}

.error {
    color: red;
    background-color: lightpink;
    font-weight: normal;
}