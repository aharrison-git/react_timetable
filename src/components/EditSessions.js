import React from 'react'
import groupData from '../data'
import GroupTypes from '../GroupTypes'

class EditSessions extends React.Component {
    constructor(props) {
        super(props)
        this.state = {class: GroupTypes[0], day:"Monday", time: "9.00", action: "Add"}
        this.handleChangeClass = this.handleChangeClass.bind(this)
        this.handleChangeDay = this.handleChangeDay.bind(this)
        this.handleChangeTime = this.handleChangeTime.bind(this)
        this.handleChangeAction = this.handleChangeAction.bind(this)
        this.handleSubmit = this.handleSubmit.bind(this)
    }

    handleChangeClass(event) {
        this.setState({class: event.target.value})
        console.log("target: " + event.target.value)
    }

    handleChangeDay(event) {
        this.setState({day: event.target.value})
        console.log("target: " + event.target.value)
    }

    handleChangeTime(event) {
        this.setState({time: event.target.value})
        console.log("target: " + event.target.value)
    }

    handleChangeAction(event) {
        this.setState({action: event.target.value})
        console.log("target: " + event.target.value)
    }

    handleSubmit(event) {
        alert("class: " + this.state.class + ", day: " + this.state.day + ", time: " + this.state.time + ", action: " + this.state.action)
        event.preventDefault()
        this.props.parentCallback(this.state)
    }

    render() {
        const classOptions = GroupTypes.map((name) => 
            <option value={name}>{name}</option>)
        const days = ["Monday", "Tuesday", "Wednesday"].map((day) =>
            <option value={day}>{day}</option>)
        const times = ["9.00", "9.45", "10.30", "11.15"].map((time) =>
            <option value={time}>{time}</option>)
        const actions = ["Add", "Delete"].map((action) =>
            <option value={action}>{action}</option>)

        return(
            <div>
                <p>Edit Sessions</p>
                
                <form class="editform" onSubmit={this.handleSubmit}>
                    <label>
                        Select Class:
                        <select value={this.state.class} onChange={this.handleChangeClass}>            
                            {classOptions}
                        </select>
                    </label>
                    <br/>
                    <label>
                        Select Day:
                        <select value={this.state.day} onChange={this.handleChangeDay}>            
                            {days}
                        </select>
                    </label>
                    <br/>
                    <label>
                        Select Time:
                        <select value={this.state.time} onChange={this.handleChangeTime}>            
                            {times}
                        </select>
                    </label>
                    <br/>
                    <label>
                        Select Action:
                        <select value={this.state.action} onChange={this.handleChangeAction}>            
                            {actions}
                        </select>
                    </label>
                    <br/>
                    
                    <input type="submit" value="Apply" />
                </form>
            </div>
        )
    }
}
//<option value="grapefruit">Grapefruit</option>
//<option value="lime">Lime</option>
//<option value="coconut">Coconut</option>
//<option value="mango">Mango</option>
export default EditSessions
