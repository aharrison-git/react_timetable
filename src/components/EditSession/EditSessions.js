import React from 'react'
import './EditSessions.css'
import groupData from '../../static/data'
import groupTypes from '../../static/GroupTypes'
import sessionTimes from '../../static/sessionTimes'
import daysOpen from '../../static/days'

class EditSessions extends React.Component {
    constructor(props) {
        super(props)
        this.state = {data: groupData, class: groupTypes[0], day:"Monday", time: "9.00", action: "Add"}
        this.handleChangeSelect = this.handleChangeSelect.bind(this)
        this.handleSubmit = this.handleSubmit.bind(this)
        this.daysOfTheWeekObject = daysOpen.map((element, index) => ({key: element, value: index}) )
        this.sessionTimes = sessionTimes.map((session) => session.start + " - " + session.end)
        this.sessionTimesObject = this.sessionTimes.map((element, index) => ({key: element, value: index}))
        this.getCode = this.getCode.bind(this)
        this.getSessionCodeString = this.getSessionCodeString.bind(this)
        this.updateData = this.updateData.bind(this)
    }

    getCode(value, objectArray) {
        for (var i=0 ; i<objectArray.length ; i++) {
            if (objectArray[i].key === value) {
                return objectArray[i].value
            }
        } 
        return null
    }


    getSessionCodeString(time, day) {
        const dayCode = this.getCode(this.state.day, this.daysOfTheWeekObject)
        const timeCode = this.getCode(this.state.time, this.sessionTimesObject)
        var sessionCode = (timeCode + 1).toString() + (dayCode + 1).toString()
        return sessionCode
    }


    updateData(name, session, action) {
        switch (action) {
            case "Add":
                var exists = 0
                for (var i=0; i<this.state.data.length; i++) {
                    if (this.state.data[i].groupName == name && this.state.data[i].session == session) {
                        exists++    
                    }
                }
                if (exists===0) {
                    var dataTemp = this.state.data
                    const newObj = {session: parseInt(session), groupName: name}
                    dataTemp.push(newObj)
                }
                break

            case "Delete":
                exists = 0
                for (i=0; i<this.state.data.length; i++) {
                    console.log(this.state.data[i].groupName + " " + this.state.data[i].session)
                    if (this.state.data[i].groupName == name && this.state.data[i].session == session) {
                        dataTemp = this.state.data.splice(i, 1)
                        break
                    }
                }
                break

            default:
                break
        }
    }

    handleChangeSelect(event) {
        this.setState({[event.target.name]: event.target.value})
    }

    handleSubmit(event) {
        const sessionCodeString = this.getSessionCodeString(this.state.time, this.state.day)
        this.updateData(this.state.class, sessionCodeString, this.state.action)
        event.preventDefault()
        this.props.parentCallback(this.state)
       }


    render() {
        const classOptions = groupTypes.map((name, index) => 
            <option key={index} value={name}>{name}</option>)
        const days = daysOpen.map((day, index) =>
            <option key={index} value={day}>{day}</option>)
        const times = this.sessionTimes.map((time, index) =>
            <option key={index} value={time}>{time}</option>)
        const actions = ["Add", "Delete"].map((action, index) =>
            <option key={index} value={action}>{action}</option>)

        return(
            <div>
                <p>Edit Sessions</p>
                
                <form className="editform" onSubmit={this.handleSubmit}>
                    <label className="cell row1 column1" for="class">Select Class</label>
                    <label className="cell row2 column1" for="day">Select Day</label>
                    <label className="cell row3 column1" for="class">Select Time</label>
                    <label className="cell row4 column1" for="class">Select Action</label>
                    <select className="cell row1 column2" value={this.state.class} name="class" onChange={this.handleChangeSelect}>{classOptions}</select>
                    <select className="cell row2 column2" value={this.state.day} name="day" onChange={this.handleChangeSelect}>{days}</select>
                    <select className="cell row3 column2" value={this.state.time} name="time" onChange={this.handleChangeSelect}>{times}</select>
                    <select className="cell row4 column2" value={this.state.action} name="action" onChange={this.handleChangeSelect}>{actions}</select>
                    <input className="cell column2 row5" type="submit" value="Apply" />
                </form>
            </div>
        )
    }
}
export default EditSessions
