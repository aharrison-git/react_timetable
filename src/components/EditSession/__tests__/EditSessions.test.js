import React from 'react'
import { create } from 'react-test-renderer'
import { shallow, mount } from 'enzyme'
import EditSessions from '../EditSessions'
//import App from '../../App'

describe("Snapshot for EditSessions component", () => {
    test("snapshot test for EditSessions component", () => {
        let tree = create(<EditSessions />)
        expect(tree.toJSON()).toMatchSnapshot()
    })
})

describe("Form renders", () => {
    test("Class select", () => {
        const wrapper = shallow(<EditSessions />)
        const classSelect = wrapper.find('[name="class"]')
        expect(classSelect).toHaveLength(1)
    })

    test("Day select", () => {
        const wrapper = shallow(<EditSessions />)
        const classSelect = wrapper.find('[name="day"]')
        expect(classSelect).toHaveLength(1)
    })

    test("Time select", () => {
        const wrapper = shallow(<EditSessions />)
        const classSelect = wrapper.find('[name="time"]')
        expect(classSelect).toHaveLength(1)
    })

    test("Action select", () => {
        const wrapper = shallow(<EditSessions />)
        const classSelect = wrapper.find('[name="action"]')
        expect(classSelect).toHaveLength(1)
    })
})

describe("Select box changes state", () => {
    test("Class select changes state", () => {
        const wrapper = shallow(<EditSessions />)
        const classSelect = wrapper.find('[name="class"]')
        classSelect.simulate('change', {target: {name: "class", value: 'Chemistry'}})
        expect(wrapper.state('class')).toEqual('Chemistry')
    })

    test("Day select changes state", () => {
        const wrapper = shallow(<EditSessions />)
        const daySelect = wrapper.find('[name="day"]')
        daySelect.simulate('change', {target: {name: "day", value: 'Tuesday'}})
        expect(wrapper.state('day')).toEqual('Tuesday')
    })

    test("Time select changes state", () => {
        const wrapper = shallow(<EditSessions />)
        const timeSelect = wrapper.find('[name="time"]')
        timeSelect.simulate('change', {target: {name: "time", value: '12.00 - 12.45'}})
        expect(wrapper.state('time')).toEqual('12.00 - 12.45')
    })

    test("Action select changes state", () => {
        const wrapper = shallow(<EditSessions />)
        const actionSelect = wrapper.find('[name="class"]')
        actionSelect.simulate('change', {target: {name: "action", value: 'Add'}})
        expect(wrapper.state('action')).toEqual('Add')
    })
})


describe("Add session", () => {
    const parentCallback = jest.fn()
    test("add new session", () => {
        const wrapper = mount(<EditSessions parentCallback={parentCallback}/>)
        
        // verify that session does not exist - session 52 = session 5, day 2
        const sessions = wrapper.state('data')
        var sessionExists = false
        sessions.forEach((session) => {
            if (session["session"] == 52) {
                sessionExists = true
            }
        })
        expect(sessionExists).toBe(false)

        // add new session
        submit_form(wrapper, 'Physics', 'Tuesday', '12.00 - 12.45', 'Add')
        console.log(wrapper.state('data'))
        const data = wrapper.state('data')
        var newSession = ''
        data.forEach((element) => {
            if (element["session"] == 52) {
                newSession = element["groupName"]
            }
        });
        expect(wrapper.state('actionSuccess')).toBe(true)
        expect(newSession).toEqual('Physics')
    })

    test("Error message when trying to add session that already exists", () => {
        const wrapper = mount(<EditSessions parentCallback={parentCallback}/>)
        
        // verify that session already exists- session 61 = session 6, day 1
        const sessions = wrapper.state('data')
        var sessionExists = false
        sessions.forEach((session) => {
            if (session["session"] == 61 && session["groupName"] == 'Chemistry') {
                sessionExists = true
            }
        })
        expect(sessionExists).toBe(true)
        
        const data_before = wrapper.state('data')

        // try to add new session
        submit_form(wrapper, 'Chemistry', 'Monday', '1.30 - 2.15', 'Add')
        const data_after = wrapper.state('data')
        expect(wrapper.state('actionSuccess')).toBe(false)
        expect(data_before).toEqual(data_after)
        expect(wrapper.find('p.error')).toHaveLength(1)
        
    })
})

describe("Delete session", () => {
    const parentCallback = jest.fn()
    test("delete existing session", () => {
        const wrapper = mount(<EditSessions parentCallback={parentCallback}/>)
        
        // verify that session exists - session 14 = session 1, day 4
        const sessions = wrapper.state('data')
        var sessionsFound = 0
        sessions.forEach((session) => {
            if (session["session"] == 14 && session["groupName"] == "English") {
                sessionsFound++
            }
        })
        expect(sessionsFound).toBe(1)

        // delete existing session
        submit_form(wrapper, 'English', 'Thursday', '9.00 - 9.45', 'Delete')
        const data = wrapper.state('data')
        sessionsFound = 0
        data.forEach((element) => {
            if (element["session"] == 14 && element["groupName"] == "English") {
                sessionsFound++
            }
        });
        expect(wrapper.state('actionSuccess')).toBe(true)
        expect(sessionsFound).toEqual(0)
    })

    test("Error message when trying to delete session that doesn't exist", () => {
        const wrapper = mount(<EditSessions parentCallback={parentCallback}/>)
        
        // verify that session doesn't exist- session 25 = session 2, day 5
        const sessions = wrapper.state('data')
        var sessionsFound = 0
        sessions.forEach((session) => {
            if (session["session"] == 25 && session["groupName"] == 'Maths') {
                sessionsFound++
            }
        })
        expect(sessionsFound).toEqual(0)
        
        const data_before = wrapper.state('data')

        // try to delete session
        submit_form(wrapper, 'Chemistry', 'Friday', '9.45 - 10.30', 'Delete')
        const data_after = wrapper.state('data')
        expect(wrapper.state('actionSuccess')).toBe(false)
        expect(data_before).toEqual(data_after)
        expect(wrapper.find('p.error')).toHaveLength(1)
        
    })
})

function submit_form(wrapper, subject, day, time, action) {
    const classSelect = wrapper.find('[name="class"]')
    const daySelect = wrapper.find('[name="day"]')
    const timeSelect = wrapper.find('[name="time"]')
    const actionSelect = wrapper.find('[name="action"]')
    const form = wrapper.find('.editform')

    classSelect.simulate('change', {target: {name: 'class', value: subject}} )
    daySelect.simulate('change', {target: {name: 'day', value: day}})
    timeSelect.simulate('change', {target: {name: 'time', value: time}})
    actionSelect.simulate('change', {target: {name: 'action', value: action}})
    form.simulate('submit', {preventDefault: () => {} })
        
}
