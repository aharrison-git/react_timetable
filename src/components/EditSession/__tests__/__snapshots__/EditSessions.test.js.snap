// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot for EditSessions component snapshot test for EditSessions component 1`] = `
<div>
  <p>
    Edit Sessions
  </p>
  <form
    className="editform"
    onSubmit={[Function]}
  >
    <label
      className="cell row1 column1"
      htmlFor="class"
    >
      Select Class
    </label>
    <label
      className="cell row2 column1"
      htmlFor="day"
    >
      Select Day
    </label>
    <label
      className="cell row3 column1"
      htmlFor="time"
    >
      Select Time
    </label>
    <label
      className="cell row4 column1"
      htmlFor="action"
    >
      Select Action
    </label>
    <select
      className="cell row1 column2"
      name="class"
      onChange={[Function]}
      value="English"
    >
      <option
        value="English"
      >
        English
      </option>
      <option
        value="Maths"
      >
        Maths
      </option>
      <option
        value="Physics"
      >
        Physics
      </option>
      <option
        value="Chemistry"
      >
        Chemistry
      </option>
      <option
        value="Music"
      >
        Music
      </option>
    </select>
    <select
      className="cell row2 column2"
      name="day"
      onChange={[Function]}
      value="Monday"
    >
      <option
        value="Monday"
      >
        Monday
      </option>
      <option
        value="Tuesday"
      >
        Tuesday
      </option>
      <option
        value="Wednesday"
      >
        Wednesday
      </option>
      <option
        value="Thursday"
      >
        Thursday
      </option>
      <option
        value="Friday"
      >
        Friday
      </option>
    </select>
    <select
      className="cell row3 column2"
      name="time"
      onChange={[Function]}
      value="9.00"
    >
      <option
        value="9.00 - 9.45"
      >
        9.00 - 9.45
      </option>
      <option
        value="9.45 - 10.30"
      >
        9.45 - 10.30
      </option>
      <option
        value="10.30 - 11.15"
      >
        10.30 - 11.15
      </option>
      <option
        value="11.15 - 12.00"
      >
        11.15 - 12.00
      </option>
      <option
        value="12.00 - 12.45"
      >
        12.00 - 12.45
      </option>
      <option
        value="1.30 - 2.15"
      >
        1.30 - 2.15
      </option>
      <option
        value="2.15 - 3.00"
      >
        2.15 - 3.00
      </option>
    </select>
    <select
      className="cell row4 column2"
      name="action"
      onChange={[Function]}
      value="Add"
    >
      <option
        value="Add"
      >
        Add
      </option>
      <option
        value="Delete"
      >
        Delete
      </option>
    </select>
    <input
      className="cell column2 row5"
      type="submit"
      value="Apply"
    />
  </form>
  <p
    className="success"
  >
    
  </p>
</div>
`;
